doc/compiler/osr/OSR.md

```diff
@@ -10,36 +10,241 @@ is available at https://www.apache.org/licenses/LICENSE-2.0.
 
-At an OSR point, the VM state can be correctly reconstructed and is consistent with the VM state needed by the interpreter to resume the execution at a given bytecode index.
+A VM event can be detected at certain points during execution of a JITed code. These points are called **OSR yield points**. JIT and VM agree on the set of the OSR yield points.
 
-To achieve that, bookkeeping is required at all possible OSR points during ILGen. Due to the cost and complexity of bookkeeping, OSR transition is limited to some well-defined points, which are calls, asyncchecks and monitor enters, etc.
+Usually, we would like the VM event to be discovered relatively soon after it took place. Garbage collection (GC) points meet such criteria.
+Therefore, depending on the event, OSR yield points usually include all or a subset of GC points.
+The reason why we would like to limit the set of the OSR yield points, if possible, is the cost of the bookkeeping associated with each such point as well as the restrictions that they impose on the optimizer.
 
-Being part of the program, OSR points like calls, asyncchecks and monitor enters can be removed or optimized and no longer be identified as OSR points. However, OSR transition may still be supported at that position.
+The actual OSR happens at **OSR Induction points**. They can happen at the time of the yield or later, as descitbed in the following sections.
 
-# potentialOSRPointHelper
-potentialOSRPointHelper is a call that serves as a marker of places where OSR transition is supported. The target bytecode index to be executed in the interpreter can be obtained by adding an offset (an information stored in the call node, the size of OSR point's bytecode in post-execution OSR) to the bytecode index of the helper call node.
+Therefore, a high level sequience of events can be described as following:
 
-The helper is not a must for OSR, but it can help locating the OSR points that are available at ILGen but disappear due to later transformations. It works as a bookkeeping tool, to track places that are OSR possible.
+    VM event => OSR yield => OSR induction => OSR transition
 
-# osrFearPointHelper
-This helper marks a place that has been optimized with runtime assumptions and requires protection of OSR mechnism. This means an OSR point is required between the invalidation point and the fear point.
 
-Similar to potentialOSRPointHelper, this helper is not a must for optimizations requiring OSR mechnism. It works as a tool to help locating the actual fear (being part of the program that may be removed or optimized to something else).
+# Voluntary and Involuntary OSR
+
+The difference between voluntary vs. involuntary OSR is whether it is induced by the JIT or the VM. In the case of involuntary OSR, the transition is induced by the VM by throwing an exception from within the VM code. Therefore, any OSR yield point can become an OSR induction point.
```

Comment:
I think this is worth clarifying. I changed:
"The difference between voluntary vs. involuntary OSR is whether it is induced by the JIT or the VM. In the case of involuntary OSR, the transition is induced by the VM _by forcibly transferring control from the VM code to an OSR handler in the compiled code. Such possible transfer of control is modelled as exception control flow in the IL because it behaves similarly._ Therefore, any OSR yield point can become an OSR induction point."